src/timeout.c

```diff
@@ -164,12 +164,19 @@ void handleBlockedClientsTimeout(void) {
 int getTimeoutFromObjectOrReply(client *c, robj *object, mstime_t *timeout, int unit) {
     long long tval;
     long double ftval;
+    mstime_t now = mstime();
 
     if (unit == UNIT_SECONDS) {
         if (getLongDoubleFromObjectOrReply(c,object,&ftval,
             "timeout is not a float or out of range") != C_OK)
             return C_ERR;
-        tval = (long long) (ftval * 1000.0);
+
+        ftval *= 1000.0;  /* seconds => millisec */
+        if (ftval > LLONG_MAX) {
+            addReplyError(c, "timeout is out of range");
+            return C_ERR;
+        }
+        tval = (long long) ftval;
```

Comment:
```suggestion
        if (ftval < 0) { /* Is this possible? Should we check it? */
            addReplyError(c,"timeout is negative");
            return C_ERR;
        }

        tval = (long long) ftval;
        if (tval > LLONG_MAX / 1000) {
            addReplyError(c,"timeout is out of range");
            return C_ERR;
        }
        tval *= 1000; /* seconds => millisec */
```

I think `long double` will not overflow or ever become larger than `LLONG_MAX` as far as I know. Probably, it is better to cast value to `long long` and do overflow checks. (Note: Not sure if we need to check negative value for `ftval`, I guess it can be negative? )

Reply:
this is the intension to make the arithmetic with double because it has no overflow issue, and compare it against LLONG_MAX just before the casting. Am I missing something?

Reply:
> if (ftval < 0) { /* Is this possible? Should we check it? */
It isn't possible by `getLongDoubleFromObjectOrReply()`. Added a test that verifies negative value.

Reply:
Yes, I was mistaken. `long double` can be larger than `LLONG_MAX`. So, it makes sense to keep the code as it is.